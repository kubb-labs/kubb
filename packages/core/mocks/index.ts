import { pascalCase } from '../src/transformers/casing.ts'

import type { PluginManager } from '../src/PluginManager.ts'
import { readSync } from '../src/fs/read.ts'

export const mockedPluginManager = {
  resolveName: ({ name, type }) => {
    if (type === 'type') {
      return pascalCase(name)
    }

    return name
  },
  resolvePath: ({ baseName }) => baseName,
  logger: {
    emit(message) {
      console.log(message)
    },
    on(eventName, args) {},
    logLevel: 'info',
  },
  getFile: ({ name, extName, pluginKey }) => {
    const baseName = `${name}${extName}`
    let source = ''

    try {
      source = readSync(baseName)
    } catch (_e) {
      //
    }

    return {
      path: baseName,
      baseName,
      meta: {
        pluginKey,
      },
      source,
    }
  },
} as PluginManager
