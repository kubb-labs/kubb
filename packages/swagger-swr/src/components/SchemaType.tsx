import { Type, usePlugin } from '@kubb/react'
import { useOperationSchemas } from '@kubb/swagger/hooks'

import type { ReactNode } from 'react'
import type { PluginOptions } from '../types.ts'

type Props = {
  factory: {
    name: string
  }
}

export function SchemaType({ factory }: Props): ReactNode {
  const {
    options: { dataReturnType },
  } = usePlugin<PluginOptions>()

  const schemas = useOperationSchemas()

  const [TData, TError, TRequest, TPathParams, TQueryParams, THeaderParams, TResponse] = [
    schemas.response.name,
    schemas.errors?.map((item) => item.name).join(' | ') || 'never',
    schemas.request?.name || 'never',
    schemas.pathParams?.name || 'never',
    schemas.queryParams?.name || 'never',
    schemas.headerParams?.name || 'never',
    schemas.response.name,
  ]

  const clientType = `${factory.name}Client`

  return (
    <>
      <Type name={clientType}>{`typeof client<${TResponse}, ${TError}, ${TRequest}>`}</Type>
      <Type name={factory.name}>
        {`
        {
          data: ${TData}
          error: ${TError}
          request: ${TRequest}
          pathParams: ${TPathParams}
          queryParams: ${TQueryParams}
          headerParams: ${THeaderParams}
          response: ${dataReturnType === 'data' ? TData : `Awaited<ReturnType<${clientType}>>`}
          client: {
            parameters: Partial<Parameters<${clientType}>[0]>
            return: Awaited<ReturnType<${clientType}>>
          }
        }
        `}
      </Type>
    </>
  )
}
