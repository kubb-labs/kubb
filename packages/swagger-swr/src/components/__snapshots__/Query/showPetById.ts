type ShowPetByIdClient = typeof client<ShowPetByIdQueryResponse, never, never>
type ShowPetById = {
  data: ShowPetByIdQueryResponse
  error: never
  request: never
  pathParams: ShowPetByIdPathParams
  queryParams: never
  headerParams: never
  response: ShowPetByIdQueryResponse
  client: {
    parameters: Partial<Parameters<ShowPetByIdClient>[0]>
    return: Awaited<ReturnType<ShowPetByIdClient>>
  }
}

export function ShowPetByIdQueryOptions<TData = ShowPetById['response']>(
  petId: ShowPetByIdPathParams['petId'],
  testId: ShowPetByIdPathParams['testId'],
  options: ShowPetById['client']['parameters'] = {},
): SWRConfiguration<TData, ShowPetById['error']> {
  return {
    fetcher: async () => {
      const res = await client<TData, ShowPetById['error']>({
        method: 'get',
        url: `/pets/${petId}`,
        ...options,
      })

      return res.data
    },
  }
}
/**
 * @summary Info for a specific pet
 * @link /pets/:petId
 */
export function useShowPetById<TData = ShowPetById['response']>(
  petId: ShowPetByIdPathParams['petId'],
  testId: ShowPetByIdPathParams['testId'],
  options?: {
    query?: SWRConfiguration<TData, ShowPetById['error']>
    client?: ShowPetById['client']['parameters']
    shouldFetch?: boolean
  },
): SWRResponse<TData, ShowPetById['error']> {
  const { query: queryOptions, client: clientOptions = {}, shouldFetch = true } = options ?? {}

  const url = `/pets/${petId}`
  const query = useSWR<TData, ShowPetById['error'], typeof url | null>(shouldFetch ? url : null, {
    ...ShowPetByIdQueryOptions<TData>(petId, testId, clientOptions),
    ...queryOptions,
  })

  return query
}
