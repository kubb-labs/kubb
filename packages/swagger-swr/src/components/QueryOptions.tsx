import transformers from '@kubb/core/transformers'
import { FunctionParams, URLPath } from '@kubb/core/utils'
import { Function, usePlugin, useResolveName } from '@kubb/react'
import { useOperation, useOperationSchemas } from '@kubb/swagger/hooks'
import { getASTParams } from '@kubb/swagger/utils'

import type { HttpMethod } from '@kubb/swagger/oas'
import type { ReactNode } from 'react'
import type { PluginOptions } from '../types.ts'

type TemplateProps = {
  /**
   * Name of the function
   */
  name: string
  /**
   * Parameters/options/props that need to be used
   */
  params: string
  /**
   * Generics that needs to be added for TypeScript
   */
  generics?: string
  /**
   * ReturnType(see async for adding Promise type)
   */
  returnType?: string
  /**
   * Options for JSdocs
   */
  JSDoc?: {
    comments: string[]
  }
  client: {
    generics: string
    method: HttpMethod
    path: URLPath
    withQueryParams: boolean
    withPathParams: boolean
    withData: boolean
    withHeaders: boolean
  }
  dataReturnType: NonNullable<PluginOptions['options']['dataReturnType']>
}

function Template({ name, params, generics, returnType, JSDoc, client, dataReturnType }: TemplateProps): ReactNode {
  const clientOptions = [
    `method: "${client.method}"`,
    `url: ${client.path.template}`,
    client.withQueryParams ? 'params' : undefined,
    client.withData ? 'data' : undefined,
    client.withHeaders ? 'headers: { ...headers, ...options.headers }' : undefined,
    '...options',
  ].filter(Boolean)

  const resolvedClientOptions = `${transformers.createIndent(4)}${clientOptions.join(`,\n${transformers.createIndent(4)}`)}`

  return (
    <Function name={name} export generics={generics} returnType={returnType} params={params} JSDoc={JSDoc}>
      {`
      return {
        fetcher: async () => {
          const res = await client<${client.generics}>({
            ${resolvedClientOptions}
          })

          return ${dataReturnType === 'data' ? 'res.data' : 'res'}
        },
      }

       `}
    </Function>
  )
}

const defaultTemplates = {
  default: Template,
} as const

type Props = {
  factory: {
    name: string
  }
  dataReturnType: NonNullable<PluginOptions['options']['dataReturnType']>
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: React.ComponentType<TemplateProps>
}

export function QueryOptions({ factory, dataReturnType, Template = defaultTemplates.default }: Props): ReactNode {
  const { key: pluginKey } = usePlugin()
  const schemas = useOperationSchemas()
  const operation = useOperation()

  const name = useResolveName({
    name: `${factory.name}QueryOptions`,
    pluginKey,
  })

  const generics = new FunctionParams()
  const params = new FunctionParams()

  const clientGenerics = ['TData', `${factory.name}['error']`]
  const resultGenerics = ['TData', `${factory.name}['error']`]

  generics.add([{ type: 'TData', default: `${factory.name}['response']` }])

  params.add([
    ...getASTParams(schemas.pathParams, { typed: true }),
    {
      name: 'params',
      type: `${factory.name}['queryParams']`,
      enabled: !!schemas.queryParams?.name,
      required: false,
    },
    {
      name: 'headers',
      type: `${factory.name}['headerParams']`,
      enabled: !!schemas.headerParams?.name,
      required: false,
    },
    {
      name: 'options',
      type: `${factory.name}['client']['parameters']`,
      default: '{}',
    },
  ])

  const client = {
    withQueryParams: !!schemas.queryParams?.name,
    withData: !!schemas.request?.name,
    withPathParams: !!schemas.pathParams?.name,
    withHeaders: !!schemas.headerParams?.name,
    method: operation.method,
    path: new URLPath(operation.path),
    generics: clientGenerics.join(', '),
  }

  return (
    <Template
      name={name}
      params={params.toString()}
      generics={generics.toString()}
      returnType={`SWRConfiguration<${resultGenerics.join(', ')}>`}
      client={client}
      dataReturnType={dataReturnType}
    />
  )
}

QueryOptions.templates = defaultTemplates
