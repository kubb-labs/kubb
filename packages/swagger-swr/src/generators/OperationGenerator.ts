import { getRelativePath } from '@kubb/core/utils'
import { OperationGenerator as Generator, resolve } from '@kubb/swagger'
import { pluginKey as swaggerTypescriptPluginKey, resolve as resolveSwaggerTypescript } from '@kubb/swagger-ts'

import { QueryBuilder } from '../builders/QueryBuilder.tsx'

import type { KubbFile } from '@kubb/core'
import type { Operation, OperationSchema, OperationSchemas, Resolver } from '@kubb/swagger'
import type { FileMeta, PluginOptions } from '../types.ts'

type Options = {
  clientImportPath?: NonNullable<PluginOptions['options']['clientImportPath']>
  dataReturnType: NonNullable<PluginOptions['options']['dataReturnType']>
}

export class OperationGenerator extends Generator<Options, PluginOptions> {
  resolve(operation: Operation): Resolver {
    const { pluginManager, plugin } = this.context

    const name = pluginManager.resolveName({ name: `use ${operation.getOperationId()}`, pluginKey: plugin.key })

    return resolve({
      name,
      operation,
      resolveName: pluginManager.resolveName,
      resolvePath: pluginManager.resolvePath,
      pluginKey: plugin.key,
    })
  }

  resolveType(operation: Operation): Resolver {
    const { pluginManager } = this.context

    return resolveSwaggerTypescript({
      operation,
      resolveName: pluginManager.resolveName,
      resolvePath: pluginManager.resolvePath,
    })
  }

  resolveError(operation: Operation, statusCode: number): Resolver {
    const { pluginManager } = this.context

    const name = pluginManager.resolveName({ name: `${operation.getOperationId()} ${statusCode}`, pluginKey: swaggerTypescriptPluginKey })

    return resolveSwaggerTypescript({
      name,
      operation,
      resolveName: pluginManager.resolveName,
      resolvePath: pluginManager.resolvePath,
    })
  }

  resolveErrors(operation: Operation, errors: OperationSchema[]): Resolver[] {
    return errors
      .map((item) => {
        if (item.statusCode) {
          return this.resolveError(operation, item.statusCode)
        }
        return undefined
      })
      .filter(Boolean)
  }

  async all(): Promise<KubbFile.File | null> {
    return null
  }

  async get(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    const { dataReturnType } = options
    const { pluginManager, oas, plugin } = this.context

    const hook = this.resolve(operation)
    const type = this.resolveType(operation)
    const clientImportPath = options.clientImportPath
      ? options.clientImportPath
      : '@kubb/swagger-client/client'

    let errors: Resolver[] = []

    if (schemas.errors) {
      errors = this.resolveErrors(operation, schemas.errors)
    }

    const queryBuilder = new QueryBuilder({ name: hook.name, errors, dataReturnType }, { oas, pluginManager, plugin, operation, schemas })

    const file = queryBuilder.render('query', hook.name).file

    if (!file) {
      throw new Error('No <File/> being used or File is undefined(see resolvePath/resolveName)')
    }

    return {
      path: file.path,
      baseName: file.baseName,
      source: file.source,
      imports: [
        ...(file.imports || []),
        {
          name: 'useSWR',
          path: 'swr',
        },
        {
          name: ['SWRConfiguration', 'SWRResponse'],
          path: 'swr',
          isTypeOnly: true,
        },
        {
          name: 'client',
          path: clientImportPath,
        },
        {
          name: ['ResponseConfig'],
          path: clientImportPath,
          isTypeOnly: true,
        },
        {
          name: [
            schemas.response.name,
            schemas.pathParams?.name,
            schemas.queryParams?.name,
            schemas.headerParams?.name,
            ...errors.map((error) => error.name),
          ].filter(Boolean),
          path: getRelativePath(hook.path, type.path),
          isTypeOnly: true,
        },
      ],
      meta: {
        pluginKey: plugin.key,
        tag: operation.getTags()[0]?.name,
      },
    }
  }

  async post(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    const { dataReturnType } = options
    const { pluginManager, oas, plugin } = this.context

    const hook = this.resolve(operation)
    const type = this.resolveType(operation)
    const clientImportPath = options.clientImportPath
      ? options.clientImportPath
      : '@kubb/swagger-client/client'

    let errors: Resolver[] = []

    if (schemas.errors) {
      errors = this.resolveErrors(operation, schemas.errors)
    }

    const queryBuilder = new QueryBuilder({ name: hook.name, errors, dataReturnType }, { oas, pluginManager, plugin, operation, schemas })

    const file = queryBuilder.render('mutation', hook.name).file

    if (!file) {
      throw new Error('No <File/> being used or File is undefined(see resolvePath/resolveName)')
    }

    return {
      path: file.path,
      baseName: file.baseName,
      source: file.source,
      imports: [
        ...(file.imports || []),
        {
          name: 'useSWRMutation',
          path: 'swr/mutation',
        },
        {
          name: ['SWRMutationConfiguration', 'SWRMutationResponse'],
          path: 'swr/mutation',
          isTypeOnly: true,
        },
        {
          name: 'client',
          path: clientImportPath,
        },
        {
          name: ['ResponseConfig'],
          path: clientImportPath,
          isTypeOnly: true,
        },
        {
          name: [
            schemas.request?.name,
            schemas.response.name,
            schemas.pathParams?.name,
            schemas.queryParams?.name,
            schemas.headerParams?.name,
            ...errors.map((error) => error.name),
          ].filter(Boolean),
          path: getRelativePath(hook.path, type.path),
          isTypeOnly: true,
        },
      ],
      meta: {
        pluginKey: plugin.key,
        tag: operation.getTags()[0]?.name,
      },
    }
  }

  async put(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
  async patch(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
  async delete(operation: Operation, schemas: OperationSchemas, options: Options): Promise<KubbFile.File<FileMeta> | null> {
    return this.post(operation, schemas, options)
  }
}
