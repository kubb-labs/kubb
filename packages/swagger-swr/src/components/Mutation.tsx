import transformers from '@kubb/core/transformers'
import { FunctionParams, URLPath } from '@kubb/core/utils'
import { Editor, File, Function, usePlugin } from '@kubb/react'
import { useGetOperationFile, useOperation, useOperationName, useOperationSchemas } from '@kubb/swagger/hooks'
import { getASTParams, getComments } from '@kubb/swagger/utils'
import { pluginKey as swaggerTsPluginKey } from '@kubb/swagger-ts'

import { SchemaType } from './SchemaType.tsx'

import type { HttpMethod } from '@kubb/swagger/oas'
import type { ReactNode } from 'react'
import type { FileMeta, PluginOptions } from '../types.ts'

type TemplateProps = {
  /**
   * Name of the function
   */
  name: string
  /**
   * Parameters/options/props that need to be used
   */
  params: string
  /**
   * Generics that needs to be added for TypeScript
   */
  generics?: string
  /**
   * ReturnType(see async for adding Promise type)
   */
  returnType?: string
  /**
   * Options for JSdocs
   */
  JSDoc?: {
    comments: string[]
  }
  hook: {
    name: string
    generics?: string
  }
  client: {
    method: HttpMethod
    generics: string
    withQueryParams: boolean
    withPathParams: boolean
    withData: boolean
    withHeaders: boolean
    path: URLPath
  }
  dataReturnType: NonNullable<PluginOptions['options']['dataReturnType']>
}

function Template({ name, generics, returnType, params, JSDoc, client, hook, dataReturnType }: TemplateProps): ReactNode {
  const clientOptions = [
    `method: "${client.method}"`,
    'url',
    client.withQueryParams ? 'params' : undefined,
    client.withData ? 'data' : undefined,
    client.withHeaders ? 'headers: { ...headers, ...clientOptions.headers }' : undefined,
    '...clientOptions',
  ].filter(Boolean)

  const resolvedClientOptions = `${transformers.createIndent(4)}${clientOptions.join(`,\n${transformers.createIndent(4)}`)}`
  if (client.withQueryParams) {
    return (
      <Function export name={name} generics={generics} returnType={returnType} params={params} JSDoc={JSDoc}>
        {`
         const { mutation: mutationOptions, client: clientOptions = {}, shouldFetch = true } = options ?? {}

         const url = ${client.path.template} as const
         return ${hook.name}<${hook.generics}>(
          shouldFetch ? [url, params]: null,
          async (_url${client.withData ? ', { arg: data }' : ''}) => {
            const res = await client<${client.generics}>({
              ${resolvedClientOptions}
            })

            return ${dataReturnType === 'data' ? 'res.data' : 'res'}
          },
          mutationOptions
        )
      `}
      </Function>
    )
  }
  return (
    <Function export name={name} generics={generics} returnType={returnType} params={params} JSDoc={JSDoc}>
      {`
       const { mutation: mutationOptions, client: clientOptions = {}, shouldFetch = true } = options ?? {}

       const url = ${client.path.template} as const
       return ${hook.name}<${hook.generics}>(
        shouldFetch ? url : null,
        async (_url${client.withData ? ', { arg: data }' : ''}) => {
          const res = await client<${client.generics}>({
            ${resolvedClientOptions}
          })

          return ${dataReturnType === 'data' ? 'res.data' : 'res'}
        },
        mutationOptions
      )
    `}
    </Function>
  )
}

const defaultTemplates = {
  default: Template,
} as const

type Props = {
  factory: {
    name: string
  }
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: React.ComponentType<TemplateProps>
}

export function Mutation({ factory, Template = defaultTemplates.default }: Props): ReactNode {
  const {
    options: { dataReturnType },
  } = usePlugin<PluginOptions>()
  const operation = useOperation()
  const name = useOperationName({ type: 'function' })
  const schemas = useOperationSchemas()

  const params = new FunctionParams()
  const client = {
    method: operation.method,
    path: new URLPath(operation.path),
    generics: [`${factory.name}["data"]`, `${factory.name}["error"]`, schemas.request?.name ? `${factory.name}["request"]` : ''].filter(Boolean).join(', '),
    withQueryParams: !!schemas.queryParams?.name,
    withData: !!schemas.request?.name,
    withPathParams: !!schemas.pathParams?.name,
    withHeaders: !!schemas.headerParams?.name,
  }

  const resultGenerics = [`${factory.name}["response"]`, `${factory.name}["error"]`]

  params.add([
    ...getASTParams(schemas.pathParams, { typed: true }),
    {
      name: 'params',
      type: `${factory.name}['queryParams']`,
      enabled: client.withQueryParams,
      required: false,
    },
    {
      name: 'headers',
      type: `${factory.name}['headerParams']`,
      enabled: client.withHeaders,
      required: false,
    },
    {
      name: 'options',
      required: false,
      type: `{
        mutation?: SWRMutationConfiguration<${resultGenerics.join(', ')}>,
        client?: ${factory.name}['client']['parameters'],
        shouldFetch?: boolean,
      }`,
      default: '{}',
    },
  ])

  const hook = {
    name: 'useSWRMutation',
    generics: [...resultGenerics, client.withQueryParams ? '[typeof url, typeof params] | null' : 'typeof url | null'].join(', '),
  }

  return (
    <Template
      name={name}
      JSDoc={{ comments: getComments(operation) }}
      client={client}
      hook={hook}
      params={params.toString()}
      returnType={`SWRMutationResponse<${resultGenerics.join(', ')}>`}
      dataReturnType={dataReturnType}
    />
  )
}

type FileProps = {
  /**
   * This will make it possible to override the default behaviour.
   */
  templates?: typeof defaultTemplates
}

Mutation.File = function ({ templates = defaultTemplates }: FileProps): ReactNode {
  const {
    options: {
      client: { importPath },
    },
  } = usePlugin<PluginOptions>()
  const schemas = useOperationSchemas()
  const file = useGetOperationFile()
  const fileType = useGetOperationFile({ pluginKey: swaggerTsPluginKey })
  const factoryName = useOperationName({ type: 'type' })

  const Template = templates.default
  const factory = {
    name: factoryName,
  }

  return (
    <Editor language="typescript">
      <File<FileMeta> baseName={file.baseName} path={file.path} meta={file.meta}>
        <File.Import name="useSWRMutation" path="swr/mutation" />
        <File.Import name={['SWRMutationConfiguration', 'SWRMutationResponse']} path="swr/mutation" isTypeOnly />
        <File.Import name={'client'} path={importPath} />
        <File.Import name={['ResponseConfig']} path={importPath} isTypeOnly />
        <File.Import
          name={[
            schemas.request?.name,
            schemas.response.name,
            schemas.pathParams?.name,
            schemas.queryParams?.name,
            schemas.headerParams?.name,
            ...(schemas.errors?.map((error) => error.name) || []),
          ].filter(Boolean)}
          root={file.path}
          path={fileType.path}
          isTypeOnly
        />

        <File.Source>
          <SchemaType factory={factory} />
          <Mutation Template={Template} factory={factory} />
        </File.Source>
      </File>
    </Editor>
  )
}

Mutation.templates = defaultTemplates
