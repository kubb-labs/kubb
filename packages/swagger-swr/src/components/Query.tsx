import { FunctionParams, URLPath } from '@kubb/core/utils'
import { Editor, File, Function, usePlugin, useResolveName } from '@kubb/react'
import { useGetOperationFile, useOperation, useOperationName, useOperationSchemas } from '@kubb/swagger/hooks'
import { getASTParams, getComments } from '@kubb/swagger/utils'
import { pluginKey as swaggerTsPluginKey } from '@kubb/swagger-ts'

import { QueryOptions } from './QueryOptions.tsx'
import { SchemaType } from './SchemaType.tsx'

import type { ReactNode } from 'react'
import type { FileMeta, PluginOptions } from '../types.ts'

type TemplateProps = {
  /**
   * Name of the function
   */
  name: string
  /**
   * Parameters/options/props that need to be used
   */
  params: string
  /**
   * Generics that needs to be added for TypeScript
   */
  generics?: string
  /**
   * ReturnType(see async for adding Promise type)
   */
  returnType?: string
  /**
   * Options for JSdocs
   */
  JSDoc?: {
    comments: string[]
  }
  hook: {
    name: string
    generics?: string
    queryOptions: string
  }
  client: {
    path: URLPath
    withQueryParams: boolean
  }
}

function Template({ name, generics, returnType, params, JSDoc, hook, client }: TemplateProps): ReactNode {
  if (client.withQueryParams) {
    return (
      <>
        <Function name={name} export generics={generics} returnType={returnType} params={params} JSDoc={JSDoc}>
          {`
         const { query: queryOptions, client: clientOptions = {}, shouldFetch = true } = options ?? {}

         const url = ${client.path.template}
         const query = ${hook.name}<${hook.generics}>(
          shouldFetch ? [url, params]: null,
          {
            ...${hook.queryOptions},
            ...queryOptions
          }
         )

         return query
         `}
        </Function>
      </>
    )
  }

  return (
    <>
      <Function name={name} export generics={generics} returnType={returnType} params={params} JSDoc={JSDoc}>
        {`
       const { query: queryOptions, client: clientOptions = {}, shouldFetch = true } = options ?? {}

       const url = ${client.path.template}
       const query = ${hook.name}<${hook.generics}>(
        shouldFetch ? url : null,
        {
          ...${hook.queryOptions},
          ...queryOptions
        }
       )

       return query
       `}
      </Function>
    </>
  )
}

const defaultTemplates = {
  default: Template,
} as const

type Props = {
  factory: {
    name: string
  }
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: React.ComponentType<TemplateProps>
  /**
   * This will make it possible to override the default behaviour.
   */
  QueryOptionsTemplate?: React.ComponentType<React.ComponentProps<typeof QueryOptions.templates.default>>
}

export function Query({ factory, Template = defaultTemplates.default, QueryOptionsTemplate = QueryOptions.templates.default }: Props): ReactNode {
  const {
    key: pluginKey,
    options: { dataReturnType },
  } = usePlugin<PluginOptions>()
  const operation = useOperation()
  const schemas = useOperationSchemas()
  const name = useOperationName({ type: 'function' })

  const queryOptionsName = useResolveName({
    name: `${factory.name}QueryOptions`,
    pluginKey,
  })
  const generics = new FunctionParams()
  const params = new FunctionParams()
  const queryParams = new FunctionParams()
  const client = {
    method: operation.method,
    path: new URLPath(operation.path),
    withQueryParams: !!schemas.queryParams?.name,
    withData: !!schemas.request?.name,
    withPathParams: !!schemas.pathParams?.name,
    withHeaders: !!schemas.headerParams?.name,
  }

  const resultGenerics = ['TData', `${factory.name}["error"]`]

  generics.add([{ type: 'TData', default: `${factory.name}["response"]` }])

  const queryOptionsGenerics = ['TData']

  params.add([
    ...getASTParams(schemas.pathParams, { typed: true }),
    {
      name: 'params',
      type: `${factory.name}['queryParams']`,
      enabled: client.withQueryParams,
      required: false,
    },
    {
      name: 'headers',
      type: `${factory.name}['headerParams']`,
      enabled: client.withHeaders,
      required: false,
    },
    {
      name: 'options',
      required: false,
      type: `{
        query?: SWRConfiguration<${resultGenerics.join(', ')}>,
        client?: ${factory.name}['client']['parameters'],
        shouldFetch?: boolean,
      }`,
      default: '{}',
    },
  ])

  queryParams.add([
    ...getASTParams(schemas.pathParams, { typed: false }),
    {
      name: 'params',
      enabled: client.withQueryParams,
      required: false,
    },
    {
      name: 'headers',
      enabled: client.withHeaders,
      required: false,
    },
    {
      name: 'clientOptions',
      required: false,
    },
  ])

  const hook = {
    name: 'useSWR',
    generics: [...resultGenerics, client.withQueryParams ? '[typeof url, typeof params] | null' : 'typeof url | null'].join(', '),
    queryOptions: `${queryOptionsName}<${queryOptionsGenerics.join(', ')}>(${queryParams.toString()})`,
  }

  return (
    <>
      <QueryOptions factory={factory} Template={QueryOptionsTemplate} dataReturnType={dataReturnType} />

      <Template
        name={name}
        generics={generics.toString()}
        JSDoc={{ comments: getComments(operation) }}
        client={client}
        hook={hook}
        params={params.toString()}
        returnType={`SWRResponse<${resultGenerics.join(', ')}>`}
      />
    </>
  )
}

type FileProps = {
  /**
   * This will make it possible to override the default behaviour.
   */
  templates?: {
    query: typeof defaultTemplates
    queryOptions: typeof QueryOptions.templates
  }
}

Query.File = function ({ templates }: FileProps): ReactNode {
  const {
    options: {
      client: { importPath },
    },
  } = usePlugin<PluginOptions>()
  const schemas = useOperationSchemas()
  const file = useGetOperationFile()
  const fileType = useGetOperationFile({ pluginKey: swaggerTsPluginKey })
  const factoryName = useOperationName({ type: 'type' })

  const Template = templates?.query.default || defaultTemplates.default
  const QueryOptionsTemplate = templates?.queryOptions.default || QueryOptions.templates.default
  const factory = {
    name: factoryName,
  }

  return (
    <Editor language="typescript">
      <File<FileMeta> baseName={file.baseName} path={file.path} meta={file.meta}>
        <File.Import name="useSWR" path="swr" />
        <File.Import name={['SWRConfiguration', 'SWRResponse']} path="swr" isTypeOnly />
        <File.Import name={'client'} path={importPath} />
        <File.Import name={['ResponseConfig']} path={importPath} isTypeOnly />
        <File.Import
          name={[
            schemas.request?.name,
            schemas.response.name,
            schemas.pathParams?.name,
            schemas.queryParams?.name,
            schemas.headerParams?.name,
            ...(schemas.statusCodes?.map((item) => item.name) || []),
          ].filter(Boolean)}
          root={file.path}
          path={fileType.path}
          isTypeOnly
        />

        <File.Source>
          <SchemaType factory={factory} />
          <Query factory={factory} Template={Template} QueryOptionsTemplate={QueryOptionsTemplate} />
        </File.Source>
      </File>
    </Editor>
  )
}

Query.templates = defaultTemplates
