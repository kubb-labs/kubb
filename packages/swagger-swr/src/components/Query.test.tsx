import { mockedPluginManager } from '@kubb/core/mocks'
import { camelCase, pascalCase } from '@kubb/core/transformers'
import { createRootServer } from '@kubb/react/server'
import { OasManager } from '@kubb/swagger'
import { Oas } from '@kubb/swagger/components'

import { OperationGenerator } from '../OperationGenerator.tsx'
import { Query } from './Query.tsx'
import { QueryOptions } from './QueryOptions.tsx'

import type { Plugin, ResolveNameParams } from '@kubb/core'
import type { AppContextProps } from '@kubb/react'
import type { GetOperationGeneratorOptions } from '@kubb/swagger'
import type { PluginOptions } from '../types.ts'

describe('<Query/>', async () => {
  const oas = await OasManager.parseFromConfig({
    root: './',
    output: { path: 'test', clean: true },
    input: { path: 'packages/swagger-swr/mocks/petStore.yaml' },
  })
  mockedPluginManager.resolveName = ({ type, name }: ResolveNameParams) => {
    if (type === 'file' || type === 'function') {
      return camelCase(`use ${name}`)
    }

    if (type === 'type') {
      return pascalCase(name)
    }
    return name
  }

  const options: GetOperationGeneratorOptions<OperationGenerator> = {
    dataReturnType: 'data',
    templates: {
      query: Query.templates,
      queryOptions: QueryOptions.templates,
    },
    client: {
      importPath: '@kubb/swagger-client/client',
    },
  }

  const plugin = { options } as Plugin<PluginOptions>
  const og = await new OperationGenerator(options, {
    oas,
    exclude: [],
    include: undefined,
    pluginManager: mockedPluginManager,
    plugin,
    contentType: undefined,
    override: undefined,
  })

  test('showPetById', async () => {
    const operation = oas.operation('/pets/{petId}', 'get')
    const context: AppContextProps<PluginOptions['appMeta']> = {
      meta: { pluginManager: mockedPluginManager, plugin },
    }

    const Component = () => {
      return (
        <Oas oas={oas} operations={[operation]} getOperationSchemas={(...props) => og.getSchemas(...props)}>
          <Oas.Operation operation={operation}>
            <Query.File />
          </Oas.Operation>
        </Oas>
      )
    }
    const root = createRootServer({ logger: mockedPluginManager.logger })
    const output = await root.renderToString(<Component />, context)

    expect(output).toMatchFileSnapshot('./__snapshots__/Query/showPetById.ts')
  })
})
