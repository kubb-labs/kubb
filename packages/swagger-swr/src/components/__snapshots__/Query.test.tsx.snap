// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Query/> > showPetById 1`] = `
"type ShowPetByIdClient = typeof client<ShowPetByIdQueryResponse, never, never>
type ShowPetById = {
  data: ShowPetByIdQueryResponse
  error: never
  request: never
  pathParams: ShowPetByIdPathParams
  queryParams: never
  headerParams: never
  response: ShowPetByIdQueryResponse
  client: {
    parameters: Partial<Parameters<ShowPetByIdClient>[0]>
    return: Awaited<ReturnType<ShowPetByIdClient>>
  }
}

export function ShowPetByIdQueryOptions<TData extends ShowPetById['response'] = ShowPetById['response'], TError = ShowPetById['error']>(
  petId: ShowPetByIdPathParams['petId'],
  testId: ShowPetByIdPathParams['testId'],
  options: ShowPetById['client']['parameters'] = {},
): SWRConfiguration<TData, TError> {
  return {
    fetcher: async () => {
      const res = await client<TData, TError>({
        method: 'get',
        url: \`/pets/\${petId}\`,
        ...options,
      })

      return res.data
    },
  }
}
/**
 * @summary Info for a specific pet
 * @link /pets/:petId */

export function useShowPetById<TData extends ShowPetById['response'] = ShowPetById['response'], TError = ShowPetById['error']>(
  petId: ShowPetByIdPathParams['petId'],
  testId: ShowPetByIdPathParams['testId'],
  options?: {
    query?: SWRConfiguration<TData, TError>
    client?: ShowPetById['client']['parameters']
    shouldFetch?: boolean
  },
): SWRResponse<TData, TError> {
  const { query: queryOptions, client: clientOptions = {}, shouldFetch = true } = options ?? {}

  const url = \`/pets/\${petId}\` as const
  const query = useSWR<TData, TError, typeof url | null>(shouldFetch ? url : null, {
    ...ShowPetByIdQueryOptions<TData, TError>(petId, testId, clientOptions),
    ...queryOptions,
  })

  return query
}
"
`;
